src/MonoTorrent/MonoTorrent.Client/PeerConnections/SocketConnection.cs

```diff
@@ -133,14 +134,40 @@ public Task<int> SendAsync(byte[] buffer, int offset, int count)
             SendArgs.SetBuffer(buffer, offset, count);
             SendArgs.UserToken = tcs;
 
-            if (!Socket.SendAsync(SendArgs))
+            if (!SK.SendAsync(SendArgs))
                 return Task.FromResult(SendArgs.BytesTransferred);
             return tcs.Task;
         }
 
         public void Dispose()
         {
-            Socket.Dispose ();
+            try
+            {
+                if (null != ReceiveArgs)
+                {
+                    ReceiveArgs.Completed -= HandleOperationCompleted;
+                    SendArgs.Completed -= HandleOperationCompleted;
```

Comment:
If we remove the `Completed -=` lines then the test should pass, and everything should be leak-free.

As `HandleOperationCompleted` is a static function then no additional memory would be accidentally retained if we *do not* remove the event handler. Also, if we *do not* remove it then we can be safely guaranteed that it will be invoked exactly once for every `SendAsync` or `ReceiveAsync` call. This means the `TaskCompletionSource` will always have a result and be marked as complete.